torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,9 +1160,60 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR)
+      << "No PGNCCL's watchdog heartbeat detected, so we are dumping debug info: "
+      << dump_nccl_trace();
+}
+
+void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+  // Logging with `FATAL`, after errMsg printed, it calls `std::abort()`
+  // to terminate the program execution.
+  LOG(FATAL) << errMsg;
+}
+
+void ProcessGroupNCCL::heartbeatMonitor() {
+  uint64_t heartBeatCounter = 0ULL;
+  while (true) {
+    auto heartbeat = heartbeat_;
+    if (heartbeat > heartBeatCounter) {
+      heartBeatCounter = heartbeat;
+      // Sleep for enough time so that heartBeat will at least increase.
+      std::this_thread::sleep_for(std::chrono::milliseconds(
+          kWatchdogThreadSleepMillis * heartbeatTimeoutInSec_));
```

Comment:
it feels the termination will get stuck for up to 2min which is not ideal. I wonder if we should check termination signal more frequently (like having an env var)

Reply:
This is controlled by env variable `TORCH_NCCL_HEARTBEAT_TIMEOUT` which is in Secs. So if you don't like 2 mins (60 * 2), users can set it to whatever you want.

Reply:
synced offline, we need to figure out a way to make it wakes up early for the normal case.

Reply:
@xw285cornell what is your expectation? 

We currently live with a 30 min nccl timeout, and try to move people to 10min instead.  Do you think it's important to have sub-minute precision with the heartbeat?

I worry that we don't know enough about the inner-workings of the comm monitor loop to ensure that it gives a reliable heartbeat in a tight bound.

Reply:
@wconstab i think the problem is that when a program correctly finished, they may have to wait for up to 2min for the process to clean up which is definitely not good imo